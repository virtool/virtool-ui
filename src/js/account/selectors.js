import { get } from "lodash-es";

export const getAccount = state => state.account;
export const getAccountAdministratorRole = state => get(state, "account.administrator_role", null);

export const getAccountId = state => state.account.id;
export const getAccountHandle = state => state.account.handle;
