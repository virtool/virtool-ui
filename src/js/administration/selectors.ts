import { get } from "lodash-es";

export const getAccountAdministratorRole = state => get(state, "account.administrator_role", null);
