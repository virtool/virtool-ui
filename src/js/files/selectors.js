import { keyBy, map, reject, sortBy } from "lodash-es";
import { createSelector } from "reselect";

const getFiles = state => state.files.documents;

export const getFilesById = createSelector(getFiles, files => keyBy(files, "id"));

export const getFilteredFileIds = createSelector(getFiles, list =>
    map(
        sortBy(
            reject(list, document => document.reserved),
            "uploaded_at"
        ).reverse(),
        "id"
    )
);
