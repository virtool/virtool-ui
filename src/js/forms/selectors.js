import { get } from "lodash-es";

export const getSingleFormValues = (state, formName) => get(state, ["forms", "formState", formName], null);
export const getAllFormState = state => state.forms.formState;
