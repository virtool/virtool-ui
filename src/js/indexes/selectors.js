import { find, get } from "lodash-es";
import { createSelector } from "reselect";

export const getIndexes = state => state.indexes.documents || [];

export const getActiveIndexId = createSelector(getIndexes, indexes =>
    get(find(indexes, { ready: true, has_files: true }), "id")
);
