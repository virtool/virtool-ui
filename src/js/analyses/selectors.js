import { filter, find, intersection, keyBy, map, reject, sortBy, toNumber, toString } from "lodash-es";
import { createSelector } from "reselect";
import { getMaxReadLength } from "../samples/selectors";
import { createFuse } from "../utils/utils";
import { fuseSearchKeys } from "./utils";

export const getActiveId = state => state.analyses.activeId;
export const getFilterOTUs = state => state.analyses.filterOTUs;
export const getFilterSequences = state => state.analyses.filterSequences;
export const getFilterAODP = state => state.analyses.filterAODP;
export const getHits = state => state.analyses.detail.results.hits;
export const getReadCount = state => state.analyses.detail.results.readCount;
export const getResults = state => state.analyses.detail.results;
export const getWorkflow = state => state.analyses.detail.workflow;

/**
 * Return a Fuse object for searching through results given an workflow name. The workflow will determine which keys the
 * search runs over.
 *
 */
export const getFuse = createSelector([getWorkflow, getHits], (workflow, hits) =>
    createFuse(hits, fuseSearchKeys[workflow], "id")
);

export const getFilterIds = createSelector(
    [getFilterAODP, getWorkflow, getHits, getFilterOTUs, getFilterSequences, getMaxReadLength, getReadCount],
    (aodpFilter, workflow, hits, filterOTUs, filterSequences, maxReadLength, readCount) => {
        if (workflow === "nuvs") {
            const filteredHits = filterSequences ? reject(hits, { e: undefined }) : hits;
            return map(filteredHits, "id");
        }

        if (workflow === "pathoscope_bowtie" && filterOTUs) {
            const filteredHits = reject(hits, hit => {
                return hit.pi * readCount < (hit.length * 0.8) / maxReadLength;
            });

            return map(filteredHits, "id");
        }

        if (workflow === "aodp" && aodpFilter) {
            const filteredHits = filter(hits, hit => {
                if (hit.identity > aodpFilter * 100) {
                    return hit.id;
                }
            });

            return map(filteredHits, fi => fi.id);
        }

        return map(hits, "id");
    }
);

export const getSearchIds = state => state.analyses.searchIds;

export const getSortKey = state => state.analyses.sortKey;

export const getSortIds = createSelector([getWorkflow, getHits, getSortKey], (workflow, hits, sortKey) => {
    switch (sortKey) {
        case "e":
            return map(sortBy(hits, "e"), "id");

        case "orfs":
            return map(sortBy(hits, "annotatedOrfCount").reverse(), "id");

        case "length":
            return map(sortBy(hits, "sequence.length").reverse(), "id");

        case "depth":
            return map(sortBy(hits, "depth").reverse(), "id");

        case "coverage":
            return map(sortBy(hits, "coverage").reverse(), "id");

        case "weight":
            return map(sortBy(hits, "pi").reverse(), "id");

        case "identity":
            return map(sortBy(hits, "identity").reverse(), "id");

        default:
            return map(hits, "id");
    }
});

export const getMatches = createSelector(
    [getWorkflow, getHits, getFilterIds, getSearchIds, getSortIds],
    (workflow, hits, filterIds, searchIds, sortIds) => {
        const matchIds = searchIds
            ? intersection(sortIds, filterIds, map(searchIds, workflow === "nuvs" ? toNumber : toString))
            : intersection(sortIds, filterIds);

        const keyed = keyBy(hits, "id");

        return map(matchIds, id => keyed[id]);
    }
);

export const getActiveHit = createSelector([getWorkflow, getMatches, getActiveId], (workflow, matches, activeId) => {
    if (activeId !== null) {
        const hit = find(matches, { id: activeId });

        if (hit) {
            return hit;
        }
    }

    return matches[0] || null;
});
