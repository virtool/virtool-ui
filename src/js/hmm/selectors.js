import { find, get } from "lodash-es";
import { getTermSelectorFactory } from "../utils/selectors";

export const getStateTerm = state => state.hmms.term;

export const getTerm = getTermSelectorFactory(getStateTerm);

export const getTask = state => {
    const tasks = state.tasks.documents;
    const taskId = get(state.hmms.status, "task.id");
    if (taskId && tasks.length) {
        const task = find(tasks, { id: taskId });
        return task || undefined;
    }
};
