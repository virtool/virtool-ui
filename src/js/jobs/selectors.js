import { get } from "lodash-es";
import { getTermSelectorFactory } from "../utils/selectors";

const getStateTerm = state => state.jobs.term;

export const getTerm = getTermSelectorFactory(getStateTerm);
export const getJobDetailId = state => get(state, "jobs.detail.id");
export const getLinkedJobs = state => state.jobs.linkedJobs;
