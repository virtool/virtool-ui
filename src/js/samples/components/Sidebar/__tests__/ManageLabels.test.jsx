import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { ManageLabels, mapDispatchToProps } from "../ManageLabels";
import { BrowserRouter } from "react-router-dom";
import { getSelectedSamples, getSelectedLabels, getPartiallySelectedLabels } from "../../../selectors";

const routerRenderWithProviders = ui => {
    const routerUi = <BrowserRouter> {ui} </BrowserRouter>;
    return renderWithProviders(routerUi);
};

describe("<ManageLabels>", () => {
    let props;
    let state;
    beforeEach(() => {
        state = {
            samples: {
                selected: ["foo_sample"],
                documents: [
                    {
                        name: "Foo Sample",
                        id: "foo_sample",
                        labels: [{ description: "", id: 1, name: "test", color: "#FCA5A5" }]
                    },
                    {
                        name: "Bar Sample",
                        id: "bar_sample",
                        labels: [
                            { description: "", id: 1, name: "test", color: "#FCA5A5" },
                            { description: "", id: 2, name: "test2", color: "#FCA5A5" }
                        ]
                    }
                ]
            }
        };
        props = {
            allLabels: [
                { description: "", id: 1, name: "test", color: "#FCA5A5" },
                { description: "", id: 2, name: "test2", color: "#FCA5A5" },
                { description: "", id: 3, name: "test3", color: "#FCA5A5" }
            ],
            selectedSamples: [
                {
                    name: "Foo Sample",
                    id: "foo_sample",
                    labels: [{ description: "", id: 1, name: "test", color: "#FCA5A5" }]
                }
            ],
            selectedLabels: {
                1: {
                    description: "",
                    id: 1,
                    name: "test",
                    color: "#FCA5A5",
                    allLabeled: true
                }
            },
            partiallySelectedLabels: [],
            onLabelUpdate: vi.fn()
        };
    });

    const updateProps = (props, state) => ({
        ...props,
        selectedSamples: getSelectedSamples(state),
        selectedLabels: getSelectedLabels(state),
        partiallySelectedLabels: getPartiallySelectedLabels(state)
    });

    it("should be disabled if no labels exist", () => {
        props.allLabels = [];
        const wrapper = shallow(<ManageLabels {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should display labels of one selected document", () => {
        renderWithProviders(<ManageLabels {...props} />);
        expect(screen.getByText("test")).toBeInTheDocument();
    });

    it("should display labels of two selected documents", () => {
        state.samples.selected = ["foo_sample", "bar_sample"];
        renderWithProviders(<ManageLabels {...updateProps(props, state)} />);
        expect(screen.getByText(`test2`)).toBeInTheDocument();
    });

    it("should call onLabelUpdate with selectedSamples, selectedLabels, and the label to be changed", async () => {
        state.samples.selected = ["foo_sample", "bar_sample"];
        props = updateProps(props, state);
        routerRenderWithProviders(<ManageLabels {...props} />);

        await userEvent.click(screen.getByRole("button", { name: "select labels" }));
        await userEvent.click(screen.getByRole("button", { name: "test3" }));
        expect(props.onLabelUpdate).toHaveBeenCalledWith(props.selectedSamples, props.selectedLabels, 3);
    });
});

describe("mapDispatchToProps", () => {
    let selectedSamples;
    let selectedLabels;
    let dispatch;
    let onLabelUpdate;

    beforeEach(() => {
        selectedSamples = [
            {
                name: "Foo Sample",
                id: "foo_sample",
                labels: [{ description: "", id: 1, name: "test", color: "#FCA5A5" }]
            },
            {
                name: "Bar Sample",
                id: "bar_sample",
                labels: [
                    { description: "", id: 1, name: "test", color: "#FCA5A5" },
                    { description: "", id: 2, name: "test2", color: "#FCA5A5" }
                ]
            }
        ];
        selectedLabels = {
            1: {
                description: "",
                id: 1,
                name: "test",
                color: "#FCA5A5",
                allLabeled: true
            },
            2: { description: "", id: 2, name: "test2", color: "#FCA5A5", allLabeled: false }
        };
        dispatch = vi.fn();
        onLabelUpdate = mapDispatchToProps(dispatch).onLabelUpdate;
    });

    it("should add new label when label is unselected", () => {
        onLabelUpdate(selectedSamples, selectedLabels, 3);
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(dispatch).toHaveBeenCalledWith({
            payload: { sampleId: "foo_sample", update: { labels: [1, 3] } },
            type: "UPDATE_SAMPLE_REQUESTED"
        });
        expect(dispatch).toHaveBeenCalledWith({
            payload: { sampleId: "bar_sample", update: { labels: [1, 2, 3] } },
            type: "UPDATE_SAMPLE_REQUESTED"
        });
    });

    it("should add new label when label is partly selected", () => {
        onLabelUpdate(selectedSamples, selectedLabels, 2);
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(dispatch).toHaveBeenCalledWith({
            payload: { sampleId: "foo_sample", update: { labels: [1, 2] } },
            type: "UPDATE_SAMPLE_REQUESTED"
        });
    });
    it("remove label from all samples when it is selected for all", () => {
        onLabelUpdate(selectedSamples, selectedLabels, 1);
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(dispatch).toHaveBeenCalledWith({
            payload: { sampleId: "foo_sample", update: { labels: [] } },
            type: "UPDATE_SAMPLE_REQUESTED"
        });
        expect(dispatch).toHaveBeenCalledWith({
            payload: { sampleId: "bar_sample", update: { labels: [2] } },
            type: "UPDATE_SAMPLE_REQUESTED"
        });
    });
});
