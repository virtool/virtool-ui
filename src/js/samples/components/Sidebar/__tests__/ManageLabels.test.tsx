import { screen, waitFor } from "@testing-library/react";
import React from "react";
import { beforeEach, describe, expect, it } from "vitest";
import { renderWithMemoryRouter } from "../../../../../tests/setupTests";
import ManageLabels from "../ManageLabels";

describe("<ManageLabels>", () => {
    let props;

    beforeEach(() => {
        props = {
            documents: [],
            selectedSamples: [],
            labels: [
                { color: "#C4B5FD", description: "", id: 1, name: "test" },
                { color: "#FCA5A5", description: "", id: 2, name: "label" },
                { color: "#1D4ED8", description: "", id: 3, name: "bar" },
            ],
        };
    });

    it("should be disabled if no labels exist", async () => {
        props.labels = [];
        renderWithMemoryRouter(<ManageLabels {...props} />);
        await waitFor(() => expect(screen.queryByLabelText("loading")).not.toBeInTheDocument());

        expect(screen.getByText("Create one")).toBeInTheDocument();
    });

    it("should display labels of one selected document", async () => {
        props.selectedSamples = props.documents = [
            {
                name: "Foo Sample",
                id: "foo_sample",
                labels: [{ color: "#C4B5FD", description: "", id: 1, name: "test" }],
            },
        ];
        renderWithMemoryRouter(<ManageLabels {...props} />);
        await waitFor(() => expect(screen.queryByLabelText("loading")).not.toBeInTheDocument());

        expect(screen.getByText("test")).toBeInTheDocument();
    });

    it("should display labels of two selected documents", async () => {
        props.selectedSamples = props.documents = [
            {
                name: "Foo Sample",
                id: "foo_sample",
                labels: [{ color: "#C4B5FD", description: "", id: 1, name: "test" }],
            },
            {
                name: "Sample",
                id: "sample",
                labels: [{ color: "#FCA5A5", description: "", id: 2, name: "label" }],
            },
        ];
        renderWithMemoryRouter(<ManageLabels {...props} />);
        await waitFor(() => expect(screen.queryByLabelText("loading")).not.toBeInTheDocument());

        expect(screen.getByText("test")).toBeInTheDocument();
        expect(screen.getByText("label")).toBeInTheDocument();
    });
});
