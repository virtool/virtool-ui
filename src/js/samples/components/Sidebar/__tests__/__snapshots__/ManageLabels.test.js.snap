// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageLabels> should be disabled if no labels exist 1`] = `
<ManageLabels__StyledSideBarSection>
  <SideBar__SidebarHeader>
    Manage Labels
    <SampleSidebarSelector
      manageLink="/samples/labels"
      onUpdate={[Function]}
      partiallySelectedItems={Array []}
      render={[Function]}
      sampleItems={Array []}
      selectedItems={
        Array [
          1,
        ]
      }
      selectionType="labels"
    />
  </SideBar__SidebarHeader>
  <SampleSidebarMultiselectList
    items={
      Object {
        "1": Object {
          "allLabeled": true,
          "color": "#FCA5A5",
          "description": "",
          "id": 1,
          "name": "test",
        },
      }
    }
    onUpdate={[Function]}
  />
  <ManageLabels__SampleLabelsFooter>
    No labels found. 
    <Link
      to="/samples/labels"
    >
      Create one
    </Link>
    .
  </ManageLabels__SampleLabelsFooter>
</ManageLabels__StyledSideBarSection>
`;

exports[`<ManageLabels> should render 1`] = `
<ManageLabels__StyledSideBarSection>
  <SideBar__SidebarHeader>
    Manage Labels
    <SampleSidebarSelector
      manageLink="/samples/labels"
      onUpdate={[Function]}
      partiallySelectedItems={Array []}
      render={[Function]}
      sampleItems={
        Array [
          Object {
            "color": "#FCA5A5",
            "description": "",
            "id": 1,
            "name": "test",
          },
          Object {
            "color": "#FCA5A5",
            "description": "",
            "id": 2,
            "name": "test2",
          },
          Object {
            "color": "#FCA5A5",
            "description": "",
            "id": 3,
            "name": "test3",
          },
        ]
      }
      selectedItems={
        Array [
          1,
        ]
      }
      selectionType="labels"
    />
  </SideBar__SidebarHeader>
  <SampleSidebarMultiselectList
    items={
      Object {
        "1": Object {
          "allLabeled": true,
          "color": "#FCA5A5",
          "description": "",
          "id": 1,
          "name": "test",
        },
      }
    }
    onUpdate={[Function]}
  />
  <Component />
</ManageLabels__StyledSideBarSection>
`;
